torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -4,14 +4,110 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
+import os
+import socket
+import threading
 from abc import ABC, abstractmethod
-from datetime import timedelta
-from typing import Any, Optional, Tuple
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Dict, Optional, Set, Tuple
 
 from torch.distributed import Store
 
 from .api import RendezvousHandler, RendezvousParameters
 
+
+@dataclass(eq=True, frozen=True)
+class _NodeDesc:
+    """Describes a node in the rendezvous.
+
+    Attributes:
+        fqdn:
+            The FQDN of the node.
+        pid:
+            The id of the process in which the rendezvous handler runs.
+        local_id:
+            A process-wide unique id.
+    """
+
+    fqdn: str
+    pid: int
+    local_id: int
+
+    def __repr__(self) -> str:
+        return f"{self.fqdn}_{self.pid}_{self.local_id}"
+
+
+class _NodeDescGenerator:
+    """Generates node descriptors.
+
+    A node descriptor is a combination of an FQDN, a process id, and an
+    auto-incremented integer that uniquely identifies a node in the rendezvous.
```

Comment:
Although quite an edge case in prod (mostly meant to be used in testing environments), it allows you to have more than one `DynamicRendezvousHandler` instance in the same process that belongs to the same job. Without a `local_id`, you cannot distinguish between the two instances. It will be more clear in the following PRs.